test/VaultTest.t.sol

```diff
@@ -0,0 +1,343 @@
+        console2.log("Rate appreciates: User is unable to borrow another half of original debt amount due to rate change");
+        console2.log("Rate appreciates: Twice the amount of collateral is needed to support initial debt position. There is no free collateral.");
+
+        uint requiredCollateral = vault.minimumCollateral(user);
+        assertEq(requiredCollateral, 0.5 ether);
+
+        vm.prank(user);
+        vm.expectRevert("Would become undercollateralized");
+        vault.borrow(0.1 ether);
+
+    }
+}
+
+abstract contract StateRateChange2 is StateRateChange {
+    function setUp() public override virtual {
+        super.setUp();
+
+        // user borrowed 1/2 of maxDebt earlier
+        // price appreciates -> 1 DAI is convertible for double the WETH
+        //1 expect revert on borrowing the other half (because of rate change) 
+        //2 borrow less and it works
+        priceFeed.updateAnswer(1e18 + (1e18 / 4));
+    }
+}
+
+contract StateRateChange2Test is StateRateChange2 {
+
+    function testBorrowLessAtNewRate() public {
+        console2.log("Rate appreciates: User can borrow a lesser amount of DAI than before");        
+        
+        //uint maxDebt = vault.getMaxDebt(user); 
+        uint maxDebt = vault.collateralToDebt(vault.deposits(user));
+        uint userDebt = vault.debts(user);
+
+        uint spareDebt = maxDebt - userDebt;
+
+        vm.expectEmit(true, true, false, true);
+        emit Borrow(address(usdc), user, spareDebt);
+        
+        vm.prank(user);
+        vault.borrow(spareDebt);
+        assertTrue(vault.debts(user) == userDebt + spareDebt);
+    }   
+
+}
+
+
+abstract contract StateLiquidated is StateBorrowed {
+    function setUp() public override virtual {
+        super.setUp();
+
+        // user borrows 1/2 of maxDebt earlier
+        // modify price to cause liquidation
+        // if 1 WETH is converted to less usdc, at mkt price, Vault will be unable to recover usdc lent
+        // USDC/WETH price must appreciate for liquidation (usdc has devalued against WETH)
+        priceFeed.updateAnswer(386372840000000 * 1e5);
+    }
+}
+
+contract StateLiquidatedTest is StateLiquidated {
+
+    function testLiquidationOnPriceAppreciation() public {
+        console2.log("Price appreciated significantly: user should be liquidated");
+        uint userDebt = vault.debts(user);
+        uint userDeposit = vault.deposits(user);
+        
+        vm.expectEmit(true, true, true, true);
+        emit Liquidation(address(weth), address(usdc), user, userDebt, userDeposit);
+
+        vault.liquidation(user);
+        assertTrue(vault.deposits(user) == 0);
+        assertTrue(vault.debts(user) == 0);
+    }
+
+    function testOnlyOwnerCanCallLiquidate() public {
+        console2.log("Only Owner of contract can call liquidate: onlyOwner modifier");
+        vm.prank(user);
+        vm.expectRevert("Ownable: caller is not the owner");
+        vault.liquidation(user);
```

Comment:
i didnt look at the tests too carefully, but on a quick pass they seem good.  i can look more carefully later